rig/machine_control/machine_controller.py

```diff
@@ -1739,34 +1807,169 @@ def get_machine(self, x=0, y=0, default_num_cores=18):
         chip_resource_exceptions = {}
 
         # Discover dead links and cores
-        for (x, y), p2p_route in iteritems(p2p_tables):
-            if x <= max_x and y <= max_y:
-                if p2p_route == consts.P2PTableEntry.none:
+        for x in range(width):
+            for y in range(height):
+                chip_info = chips_info.get((x, y))
+                if chip_info is None:
                     dead_chips.add((x, y))
                 else:
-                    try:
-                        num_working_cores = self.get_num_working_cores(x, y)
-                        working_links = self.get_working_links(x, y)
-
-                        if num_working_cores < default_num_cores:
-                            resource_exception = chip_resources.copy()
-                            resource_exception[Cores] = min(default_num_cores,
-                                                            num_working_cores)
-                            chip_resource_exceptions[(x, y)] = \
-                                resource_exception
-
-                        for link in set(Links) - working_links:
-                            dead_links.add((x, y, link))
-                    except SCPError:
-                        # The chip was listed in the P2P table but is not
-                        # responding. Assume it is dead anyway.
-                        dead_chips.add((x, y))
+                    if chip_info.num_cores < default_num_cores:
+                        resource_exception = chip_resources.copy()
+                        resource_exception[Cores] = min(default_num_cores,
+                                                        chip_info.num_cores)
+                        chip_resource_exceptions[(x, y)] = \
+                            resource_exception
 
-        return Machine(max_x + 1, max_y + 1,
+                    for link in set(Links) - chip_info.working_links:
+                        dead_links.add((x, y, link))
+
+        return Machine(width, height,
                        chip_resources,
                        chip_resource_exceptions,
                        dead_chips, dead_links)
 
+    @ContextMixin.use_contextual_arguments()
+    def get_resource_constraints(self, machine,
+                                 core_resource=Cores,
+                                 sdram_resource=SDRAM,
+                                 sram_resource=SRAM,
+                                 _all_chips_info=None):
+        """Returns a set of place-and-route
```

Comment:
As per PEP257:
"The docstring is a phrase ending in a period. It prescribes the function or method's effect as a command ("Do this", "Return that"), not as a description; e.g. don't write "Returns the pathname ...". "